Targeted Interoperability between Swift Testing and XCTest (#2982)
swiftlang/swift-evolution: proposals/testing/NNNN-targeted-interoperability-swift-testing-and-xctest.md

# Targeted Interoperability between Swift Testing and XCTest

- Proposal: [ST-NNNN](NNNN-xctest-interoperability.md)
- Authors: [Jerry Chen](https://github.com/jerryjrchen)
- Review Manager: TBD
- Status: **Awaiting implementation**
- Implementation: [swiftlang/swift-testing#NNNNN](https://github.com/swiftlang/swift-testing/pull/NNNNN)
- Review: ([pitch](https://forums.swift.org/...))

## Introduction

Many projects want to migrate from XCTest to Swift Testing, and may be in an
intermediate state where test helpers written using XCTest API are called from
Swift Testing. Today, the Swift Testing and XCTest libraries stand mostly
independently, which means an `XCTAssert` failure in a Swift Testing test is
silently ignored. To address this, we formally declare a set of interoperability
principles and propose updates to specific APIs that will enable users to
migrate with confidence.

## Motivation

Unfortunately, mixing an API call from one framework with a test from the other

> framework may not work as expected. As a more concrete example, if you take an

Author: How about

"Calling XCTest or Swift Testing API within a test from the opposite framework may not always work as expected"

existing test helper function written for XCTest and call it in a Swift Testing
test, it won't report the assertion failure:

```swift
func assertUnique(_ elements: [Int]) {
  if Set(elements).count != elements.count {
    XCTFail("\(elements) has non unique elements")
  }
}

// XCTest

class FooTests: XCTestCase {
    func testDups() {
        assertUnique([1, 2, 1]) // Fails as expected
    }
}

// Swift Testing

@Test func `Duplicate elements`() {
  assertUnique([1, 2, 1]) // Passes? Oh no!
}
```

Generally, we get into trouble today when ALL the following conditions are met:

- You call XCTest API in a Swift Testing test, or call Swift Testing API in a
  XCTest test
- The API doesn't function as expected in some or all cases, and
- You get no notice at build time or runtime about the malfunction

For the remainder of this proposal, we’ll describe tests which exhibit this
problem as **lossy without interop**.

If you've switched completely to Swift Testing and don't expect to use XCTest in
the future, this proposal includes a mechanism to **prevent you from
inadvertently introducing XCTest APIs to your project**, including via a testing

> library.

Reviewer: Since this is still the Motivation section, I would phrase this in terms of what the problem you're seeking to address is. In this case, I think the problem (which is secondary from the earlier motivation) is that some users may wish to enforce that they only use Swift Testing and do not mistakenly use XCTest.

Author: Makes sense. I also added our primary motivation clearly at the end of this section

## Proposed solution

- **XCTest APIs which are lossy without interop will work as expected when
  called in Swift Testing.** In addition, runtime diagnostics will be included
  to highlight opportunities to adopt newer Swift Testing equivalent APIs.

- Conversely, **Swift Testing API called in XCTest will work as expected if
  XCTest already provides similar functionality.** You should always feel
  empowered to choose Swift Testing when writing new tests or test helpers, as
  it will work properly in both types of tests.

> We don't propose supporting interoperability for APIs without risk of data loss,

Reviewer: ```suggestion
We don't propose supporting interoperability for APIs which are not lossy without interop,
```

Author: Don't not hate the double negative here, but can't think of better wording at the moment :)

because they naturally have high visibility. For example, using `throw XCTSkip`
in a Swift Testing test results in a test failure rather than a test skip,
providing a clear indication that migration is needed.

## Detailed design

### Highlight and support XCTest APIs which are lossy without interop

We propose supporting the following XCTest APIs in Swift Testing:

> - Assertions: `XCTAssert*` and [unconditional failure][] `XCTFail`

Reviewer: I realize there's not a single page in the XCTest documentation which collects all `XCTAssert` functions, but I think a close approximation of that is the [Equality and Inequality Assertions](https://developer.apple.com/documentation/xctest/equality-and-inequality-assertions) page, so you could link to that as a representative example.

- [Expected failures][], such as `XCTExpectFailure`: marking a Swift Testing
  issue in this way will generate a runtime warning issue.

> - `XCTAttachment`

Reviewer: Could link to [Adding Attachments to Tests, Activities, and Issues](https://developer.apple.com/documentation/xctest/adding-attachments-to-tests-activities-and-issues)

- [Issue handling traits][]: we will make our best effort to translate issues
  from XCTest to Swift Testing. Note that there are certain issue kinds that are

>   new to Swift Testing and not expressible from XCTest.

Author: And then include any additional details as a comment so it isn't lost.

> Note that no changes are proposed for the `XCTSkip` API, because they already
> feature prominently as a test failure to be corrected when thrown in Swift
> Testing.

Reviewer: ```suggestion
Note that no changes are proposed for the `XCTSkip` API, because they already
feature prominently as a test failure when thrown by a Swift Testing test.
```

We also propose highlighting usage of above XCTest APIs in Swift Testing:

- **Report [runtime warning issues][]** for XCTest API usage in Swift Testing.
  This **applies to assertion successes AND failures**! We want to make sure you
  can identify opportunities to modernise even if your tests currently pass.

- Opt-in **strict interop mode**, where XCTest API usage will result in
  `fatalError("Usage of XCTest API in a Swift Testing context is not

> supported")`.

Reviewer: indentation

Here are some concrete examples:

| When running a Swift Testing test... | Current         | Proposed                                     | Proposed (strict) |
| ------------------------------------ | --------------- | -------------------------------------------- | ----------------- |
| `XCTAssert` failure is a ...         | ‼️ No-op        | ❌ Test Failure and ⚠️ Runtime Warning Issue | 💥 `fatalError`   |
| `XCTAssert` success is a ...         | No-op           | ⚠️ Runtime Warning Issue                     | 💥 `fatalError`   |
| `throw XCTSkip` is a ...             | ❌ Test Failure | ❌ Test Failure                              | ❌ Test Failure   |

### Targeted support for Swift Testing APIs with XCTest API equivalents

We propose supporting the following Swift Testing APIs in XCTest:

- `#expect` and `#require`
  - Includes [exit testing][]
- `withKnownIssue`: marking an XCTest issue in this way will generate a runtime
  warning issue. In strict interop mode, this becomes a `fatalError`.
- Attachments

> - [Test cancellation][] (links to pitch)

Reviewer: ```suggestion
- [Test cancellation][] (currently pitched)
```

We think developers will find utility in using Swift Testing APIs in XCTest. For
example, you can replace `XCTAssert` with `#expect` in your XCTest tests and
immediately get the benefits of the newer, more ergonomic API. In the meantime,
you can incrementally migrate the rest of your test infrastructure to use Swift
Testing at your own pace.

Present and future Swift Testing APIs will be supported in XCTest if the
XCTest API _already_ provides similar functionality.

> - For example, we support the proposed Swift Testing [test cancellation][]
>   feature in XCTest since it is analogous to `XCTSkip`.

Reviewer: ```suggestion
- For example, if the recently-pitched [test cancellation][] feature in Swift Testing were to be accepted in the future,
  this proposal means that it would be supported in an XCTest test since it's analogous to `XCTSkip`.
```

- On the other hand, [Traits][] are a powerful Swift Testing feature, and
  include the ability to [add tags][tags] to organise tests. Even though XCTest
  does not interact with tags, **this is beyond the scope of interoperability**
  because XCTest doesn't have existing “tag-like” behaviour to map onto.

Here are some concrete examples:

| When running a XCTest test...                | Current         | Proposed                 | Proposed (strict) |
| -------------------------------------------- | --------------- | ------------------------ | ----------------- |
| `#expect` failure is a ...                   | ‼️ No-op        | ❌ Test Failure          | ❌ Test Failure   |
| `#expect` success is a ...                   | No-op           | No-op                    | No-op             |
| `withKnownIssue` wrapping `XCTFail` is a ... | ❌ Test Failure | ⚠️ Runtime Warning Issue | 💥 `fatalError`   |

### Interoperability Modes

- **Warning-only**: This is for projects which do not want to see new test

>   failures surfaced due to interoperability.

Reviewer: Can this bullet state what it _does_, i.e. that it records a warning-severity issue for test failure _and_ a second warning issue indicating the XCTest usage?

| Interop Mode | Issue behaviour across framework boundary                         | `SWIFT_TESTING_XCTEST_INTEROP_MODE`            |
| ------------ | ----------------------------------------------------------------- | ---------------------------------------------- |
| Warning-only | XCTest API: ⚠️ Runtime Warning Issue                              | `warning-only`                                 |
| Permissive   | XCTest API: ⚠️ Runtime Warning Issue. All Issues: ❌ Test Failure | `permissive`, or empty value, or invalid value |
| Strict       | XCTest API: 💥 `fatalError`. Swift Testing API: ❌ Test Failure   | `strict`                                       |

### Phased Rollout

When interoperability is first available, "permissive" will be the default
interop mode enabled for new projects. In a future release, "strict" will become
the default interop mode.

## Source compatibility

As the main goal of interoperability is to change behaviour, this proposal will
lead to situations where previously "passing" test code now starts showing
failures. We believe this should be a net positive if it can highlight actual
bugs you would have missed previously.

> You can use `SWIFT_TESTING_XCTEST_INTEROP_MODE=off` in the short-term to revert

Reviewer: Since `off` mode has been removed from the proposal, we may need to adjust this recommendation and suggest `warning-only`.

Author: Good catch!

- After new API is added to Swift Testing in future, will need to evaluate for
  interoperability with XCTest. Once strict mode is the default, we will no
  longer include interoperability for new Swift Testing features.

## Alternatives considered

### Arguments against interoperability

Frameworks that operate independently generally have no expectation of
interoperability, and XCTest and Swift Testing are currently set up this way.
Indeed, the end goal is not necessarily to support using XCTest APIs
indefinitely within Swift Testing. Adding interoperability can be interpreted as
going against that goal, and can enable users to delay migrating completely to
Swift Testing.

However, we believe the benefits of fixing lossy without interop APIs and
helping users catch more bugs are too important to pass up. We've also included
a plan to increase the strictness of the interoperability mode over time, which
should make it clear that this is not intended to be a permanent measure.

### Opt-out of interoperability

In a similar vein, we considered `SWIFT_TESTING_XCTEST_INTEROP_MODE=off` as a
way to completely turn off interoperability. Some projects may additionally have

> issue handling trait that promote warnings to errors, which means that

Reviewer: ```suggestion
an issue handling trait that promotes warnings to errors, which means that
```

warning-only mode could still cause test failures.

However, in the scenario above, we think users who set up tests to elevate
warnings as errors would be interested in increased visibility of testing issues
surfaced by interop. We're open to feedback about other scenarios where a
"interop off" mode would be preferred.

### Strict interop mode as the default

We believe that for projects using only Swift Testing, strict interop mode is
the best choice. Making this the default would also send the clearest signal
that we want users to migrate to Swift Testing.

However, we are especially sensitive to use cases that depend upon the currently
lossy without interop APIs, and decided to prioritise the current default as a

> good balance between notifying users yet not breaking existing test suites.

Reviewer: > …yet not breaking existing test suites.

Well, since we're proposing `permissive` as the default, that _could_ break test suites — "break" in the sense of changing a test from passing to failing, even though it won't outright crash. We should probably clarify this to differentiate between the levels of impact: `strict` mode would mean the test process completely stops upon the first usage of an XCTest API in a Swift Testing tests, whereas `permissive` would record a failure but allow the test process to proceed to subsequent tests.

# Targeted Interoperability between Swift Testing and XCTest

- Proposal: [ST-NNNN](NNNN-xctest-interoperability.md)
- Authors: [Jerry Chen](https://github.com/jerryjrchen)
- Review Manager: TBD
- Status: **Awaiting implementation**
- Implementation: [swiftlang/swift-testing#NNNNN](https://github.com/swiftlang/swift-testing/pull/NNNNN)
- Review: ([pitch](https://forums.swift.org/...))

## Introduction

Many projects want to migrate from XCTest to Swift Testing, and may be in an
intermediate state where test helpers written using XCTest API are called from
Swift Testing. Today, the Swift Testing and XCTest libraries stand mostly

> independently, which means an `XCTAssert` failure in a Swift Testing test is

Reviewer 2: Consider including a link here to the documentation for `XCTAssert()`.

> silently ignored. To address this, we formally declare a set of interoperability

Reviewer 2: (And vice versa!)

Author: Updated wording here although not sure if that clarifies things enough

principles and propose changes to the handling of specific APIs that will enable
users to migrate with confidence.

## Motivation

Calling XCTest or Swift Testing API within a test from the opposite framework
may not always work as expected. As a more concrete example, if you take an
existing test helper function written for XCTest and call it in a Swift Testing
test, it won't report the assertion failure:

```swift
func assertUnique(_ elements: [Int]) {
  if Set(elements).count != elements.count {
    XCTFail("\(elements) has non unique elements")
  }
}

// XCTest

class FooTests: XCTestCase {
    func testDups() {
        assertUnique([1, 2, 1]) // Fails as expected
    }
}

// Swift Testing

@Test func `Duplicate elements`() {
  assertUnique([1, 2, 1]) // Passes? Oh no!
}
```

Generally, we get into trouble today when ALL the following conditions are met:

> - You call XCTest API in a Swift Testing test, or call Swift Testing API in a

Reviewer 2: Does this list describe the conditions under which trouble is got into, or is it describing the trouble itself? :)

Author: Mostly the latter. I tried rewording so it focuses more on today's API limitation

Testing. Furthermore, projects that have switched completely to Swift Testing
may want to go and ensure they don't inadvertently add XCTest API.

## Proposed solution

- **XCTest APIs which are lossy without interop will work as expected when
  called in Swift Testing.** In addition, runtime diagnostics will be included
  to highlight opportunities to adopt newer Swift Testing equivalent APIs.

- Conversely, **Swift Testing API called in XCTest will work as expected if
  XCTest already provides similar functionality.** You should always feel
  empowered to choose Swift Testing when writing new tests or test helpers, as
  it will work properly in both types of tests.

We don't propose supporting interoperability for APIs which are not lossy
without interop, because they naturally have high visibility. For example, using
`throw XCTSkip` in a Swift Testing test results in a test failure rather than a
test skip, providing a clear indication that migration is needed.

## Detailed design

### Highlight and support XCTest APIs which are lossy without interop

We propose supporting the following XCTest APIs in Swift Testing:

> - [Assertions][XCTest Assertions]: `XCTAssert*` and [unconditional failure][]

Reviewer 2: I just learned a new Markdown syntax!

  `XCTFail`
- [Expected failures][], such as `XCTExpectFailure`: marking a Swift Testing
  issue in this way will generate a runtime warning issue.
- [`XCTAttachment`][XCTest attachments]
- [Issue handling traits][]: we will make our best effort to translate issues
  from XCTest to Swift Testing. For issue details unique to XCTest, we will
  include them as a comment when constructing the Swift Testing issue.

> Note that no changes are proposed for the `XCTSkip` API, because they already

Reviewer 2: `XCTestAssertion` and `XCTWaiter`, I guess!

> Note that no changes are proposed for the `XCTSkip` API, because they already

Reviewer: (I assume you meant XCTest**Expectation**)

  usage occurs when running Swift Testing tests by turning those warnings into a
  `fatalError`.

Configure the interoperability mode when running tests using the
`SWIFT_TESTING_XCTEST_INTEROP_MODE` environment variable:

| Interop Mode | Issue behaviour across framework boundary                                  | `SWIFT_TESTING_XCTEST_INTEROP_MODE`            |
| ------------ | -------------------------------------------------------------------------- | ---------------------------------------------- |
| Warning-only | XCTest API: ⚠️ Runtime Warning Issue. All Issues: ⚠️ Runtime Warning Issue | `warning-only`                                 |
| Permissive   | XCTest API: ⚠️ Runtime Warning Issue. All Issues: ❌ Test Failure          | `permissive`, or empty value, or invalid value |
| Strict       | XCTest API: 💥 `fatalError`. Swift Testing API: ❌ Test Failure            | `strict`                                       |

### Phased Rollout

When interoperability is first available, "permissive" will be the default
interop mode enabled for new projects. In a future release, "strict" will become
the default interop mode.

## Source compatibility

> As the main goal of interoperability is to change behaviour, this proposal will

Reviewer 2: ```suggestion
As the main goal of interoperability is to change behavior, this proposal will
```

Alas!

Author: Will do a spelling check for 🇬🇧 -> 🇺🇸

lead to situations where previously "passing" test code now starts showing
failures. We believe this should be a net positive if it can highlight actual
bugs you would have missed previously.

You can use `SWIFT_TESTING_XCTEST_INTEROP_MODE=warning-only` in the short-term
to revert any changes to test pass/fail outcomes as a result of
interoperability.

## Integration with supporting tools

> Interoperability will be first available in future toolchain version,

Author: I'll move phased rollout down here

hypothetically named `6.X`, where permissive interop mode will be enabled for
projects. After that, a `7.Y` release would make strict interop mode the
default.

- Swift Package Manager projects: `swift-tools-version` declared in
  Package.swift will be used to determine interop mode, regardless of the
  toolchain used to run tests.

- Xcode projects: Installed toolchain version will be used to determine interop
  mode.

- Any project can use `SWIFT_TESTING_XCTEST_INTEROP_MODE` to override interop

>   mode at runtime, provided they are on toolchain version `6.X` or newer

Reviewer 2: ```suggestion
  mode at runtime.
```

No need to say "provided they are on X" as it's implied by the developer not having a time machine :D

# Targeted Interoperability between Swift Testing and XCTest

- Proposal: [ST-NNNN](NNNN-xctest-interoperability.md)
- Authors: [Jerry Chen](https://github.com/jerryjrchen)
- Review Manager: TBD
- Status: **Awaiting implementation**
- Implementation: [swiftlang/swift-testing#NNNNN](https://github.com/swiftlang/swift-testing/pull/NNNNN)
- Review: ([pitch](https://forums.swift.org/...))

## Introduction

Many projects want to migrate from XCTest to Swift Testing, and may be in an
intermediate state where test helpers written using XCTest API are called from
Swift Testing. Today, the Swift Testing and XCTest libraries stand mostly
independently, which means an [`XCTAssert`][XCTest assertions] failure in a

> Swift Testing test or an [`#expect`][Swift Testing expectations] failure in an

Reviewer 2: ```suggestion
Swift Testing test or an [`#expect()`][Swift Testing expectations] failure in an
```

You could regress test coverage if you migrate to Swift Testing without
replacing usage of lossy without interop test APIs. Furthermore, you may want to
ensure you don't inadvertently introduce new XCTest API after completing your
Swift Testing migration.

## Proposed solution

- **XCTest APIs which are lossy without interop will work as expected when
  called in Swift Testing.** In addition, runtime diagnostics will be included
  to highlight opportunities to adopt newer Swift Testing equivalent APIs.

- Conversely, **Swift Testing API called in XCTest will work as expected if
  XCTest already provides similar functionality.** You should always feel
  empowered to choose Swift Testing when writing new tests or test helpers, as
  it will work properly in both types of tests.

We don't propose supporting interoperability for APIs which are not lossy
without interop, because they naturally have high visibility. For example, using
`throw XCTSkip` in a Swift Testing test results in a test failure rather than a
test skip, providing a clear indication that migration is needed.

## Detailed design

> [!NOTE]
> This proposal refers to XCTest in the abstract. There are two different

> > implementations of XCTest: the open source [Corelibs XCTest][] and a

Author: I'll just use `swift-corelibs-xctest` for now since that's the repo name, and the readme on that project mentions a similarly named project in the same way

> This project is only guaranteed to build with the very latest commit on the Swift and swift-corelibs-foundation main branches. You may update to the latest commits using the Swift utils/update-checkout script:

> [proprietary XCTest][Xcode XCTest] that is shipped as part of Xcode. The Swift
> evolution process governs changes to the former only. Therefore,
> this proposal is targeted for Corelibs XCTest.

[Corelibs XCTest]: https://github.com/swiftlang/swift-corelibs-xctest
[Xcode XCTest]: https://developer.apple.com/documentation/xctest

### Highlight and support XCTest APIs which are lossy without interop

We propose supporting the following XCTest APIs in Swift Testing:

- [Assertions][XCTest Assertions]: `XCTAssert*` and [unconditional failure][]
  `XCTFail`
- [Expected failures][], such as `XCTExpectFailure`: marking a Swift Testing
  issue in this way will generate a runtime warning issue.
- [`XCTAttachment`][XCTest attachments]

> - [Issue handling traits][]: we will make our best effort to translate issues

Reviewer 2: Is this point about _traits_ or about issue handling in general? (Since we're not translating traits.)

| `throw XCTSkip` is a ...             | ❌ Test Failure           | ❌ Test Failure                              | ❌ Test Failure   |

### Targeted support for Swift Testing APIs with XCTest API equivalents

We propose supporting the following Swift Testing APIs in XCTest:

- [`#expect` and `#require`][Swift Testing expectations]
  - Includes [`#expect(throws:)`][testing for errors]
  - Includes [exit testing][]
  - Includes directly recording an issue with [`Issue.record()`][issues]
- `withKnownIssue`: marking an XCTest issue in this way will generate a runtime
  warning issue. In strict interop mode, this becomes a `fatalError`.
- [Test cancellation][]

We think developers will find utility in using Swift Testing APIs in XCTest. For
example, you can replace `XCTAssert` with `#expect` in your XCTest tests and
immediately get the benefits of the newer, more ergonomic API. In the meantime,
you can incrementally migrate the rest of your test infrastructure to use Swift
Testing at your own pace.

Present and future Swift Testing APIs will be supported in XCTest if the
XCTest API _already_ provides similar functionality.

> For example, the recently-pitched [test cancellation][] feature in Swift Testing

Author: Do'h I fixed the link elsewhere but forgot to update it here
